File: quant1x/level1/security_count.rs
#![allow(dead_code)]
use super::sequence_id;
use crate::level1::commands::*;
use crate::level1::protocol::{Request, RequestHeader, Response, ResponseHeader};
use crate::std::BinaryStream;

// Request builder for SECURITY_COUNT
#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct SecurityCountRequest {
    header: RequestHeader,
    pub market: u16,
}

impl Request for SecurityCountRequest {
    fn header(&self) -> &RequestHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut RequestHeader {
        &mut self.header
    }

    fn serialize_payload(&mut self) -> Vec<u8> {
        let mut buf = BinaryStream::new();
        // payload
        buf.push_u16(self.market);
        // padding bytes: 75c73301
        buf.push_byte_array(&[0x75, 0xc7, 0x33, 0x01]);
        buf.data().clone()
    }

    fn payload_string(&self) -> String {
        format!("SecurityCountRequest{{market:{}}}", self.market)
    }
}

impl SecurityCountRequest {
    pub fn new(market: u16) -> Self {
        let mut header = RequestHeader::new();
        header.zip_flag = crate::level1::protocol::zlib_flag::UNCOMPRESSED;
        header.seq_id = sequence_id();
        header.packet_type = 0x01;
        header.method = SECURITY_COUNT;

        SecurityCountRequest { header, market }
    }
}

#[allow(dead_code)]
#[derive(Debug, Clone)]
pub struct SecurityCountResponse {
    header: ResponseHeader,
    pub count: usize,
}

impl Response for SecurityCountResponse {
    fn header(&self) -> &ResponseHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut ResponseHeader {
        &mut self.header
    }

    fn deserialize_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        if data.len() < 2 {
            return Ok(());
        }
        let mut bs = BinaryStream::from_vec(data.to_vec());
        let c = bs.get_u16()?;
        self.count = c as usize;
        Ok(())
    }

    fn body_string(&self) -> String {
        format!("SecurityCountResponse{{count:{}}}", self.count)
    }
}

#[allow(dead_code)]
impl SecurityCountResponse {
    pub fn new() -> Self {
        Self {
            header: ResponseHeader::new(),
            count: 0,
        }
    }
}

/// Fetch security count for a market from level1 server.
pub fn fetch_security_count(market: u16) -> Option<SecurityCountResponse> {
    match crate::level1::client::get_std_conn() {
        Ok(mut pooled) => {
            let mut req = SecurityCountRequest::new(market);
            let mut resp = SecurityCountResponse::new();
            match crate::level1::process(pooled.stream(), &mut req, &mut resp) {
                Ok(_) => {
                    log::info!(
                        "level1::security_count - market={} count={}",
                        market,
                        resp.count
                    );
                    Some(resp)
                }
                Err(e) => {
                    log::error!(
                        "level1 process error for security_count market={}: {}",
                        market,
                        e.to_string()
                    );
                    None
                }
            }
        }
        Err(e) => {
            log::error!(
                "failed to acquire level1 client for security_count market={}: {}",
                market,
                e
            );
            None
        }
    }
}

File: quant1x/level1/security_count.py
# -*- coding: UTF-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from __future__ import annotations

import struct

from quant1x.data.meta import Timestamp
from quant1x.level1.protocol import (
    FLAG_UNCOMPRESSED,
    COMMAND_SECURITY_COUNT,
    sequence_id,
)

class SecurityCountRequest:
    """
    证券统计请求
    """
    def __init__(self, market: int = 0):
        self.zip_flag = FLAG_UNCOMPRESSED
        self.seq_id = sequence_id()
        self.packet_type = 0x01
        self.body_wire_len = 0
        self.body_raw_len = 0
        self.method = COMMAND_SECURITY_COUNT
        self.market = market
        yyyymmdd = Timestamp.now().yyyymmdd() # 当日日期，格式为 YYYYMMDD
        self.padding = struct.pack('<I', yyyymmdd) # 4字节日期填充

    def serialize(self) -> bytes:
        # Body: Market(2) + Padding(4) = 6 bytes
        # PkgLen = Body + 2 = 8
        self.body_wire_len = 2 + 2 + len(self.padding)
        self.body_raw_len = self.body_wire_len
        
        header = struct.pack('<B I B H H H', self.zip_flag, self.seq_id, self.packet_type, self.body_wire_len, self.body_raw_len, self.method)
        
        body = struct.pack('<H', self.market) + self.padding
        return header + body

class SecurityCountResponse:
    """
    证券统计响应
    """
    def __init__(self):
        self.count = 0

    def deserialize(self, data: bytes) -> None:
        if len(data) >= 2:
            self.count = struct.unpack('<H', data[:2])[0]
